lgc/state/PalMetadata.cpp

```diff
@@ -969,3 +920,178 @@ void PalMetadata::eraseFragmentInputInfo() {
+  }
+  return CallingConv::AMDGPU_CS;
+}
+
+// =====================================================================================================================
+// Returns the offset of the first user data register for the given calling convention.
+//
+// @param callingConv : The calling convention
+unsigned PalMetadata::getFirstUserDataReg(unsigned callingConv) {
+  static const std::map<unsigned, unsigned> shaderTable = {{CallingConv::AMDGPU_LS, mmSPI_SHADER_USER_DATA_LS_0},
+                                                           {CallingConv::AMDGPU_HS, mmSPI_SHADER_USER_DATA_HS_0},
+                                                           {CallingConv::AMDGPU_ES, mmSPI_SHADER_USER_DATA_ES_0},
+                                                           {CallingConv::AMDGPU_GS, mmSPI_SHADER_USER_DATA_GS_0},
+                                                           {CallingConv::AMDGPU_VS, mmSPI_SHADER_USER_DATA_VS_0}};
+  static const std::map<unsigned, unsigned> shaderTableGfx9 = {{CallingConv::AMDGPU_LS, mmSPI_SHADER_USER_DATA_LS_0},
+                                                               {CallingConv::AMDGPU_HS, mmSPI_SHADER_USER_DATA_LS_0},
+                                                               {CallingConv::AMDGPU_ES, mmSPI_SHADER_USER_DATA_ES_0},
+                                                               {CallingConv::AMDGPU_GS, mmSPI_SHADER_USER_DATA_ES_0},
+                                                               {CallingConv::AMDGPU_VS, mmSPI_SHADER_USER_DATA_VS_0}};
+
+  bool isGfx9 = m_pipelineState->getTargetInfo().getGfxIpVersion().major == 9;
+  const std::map<unsigned, unsigned> &currentShaderTable = (isGfx9 ? shaderTableGfx9 : shaderTable);
+  return currentShaderTable.at(callingConv);
+}
+
+// =====================================================================================================================
+// Returns to number of SGPRs before the SGPRs used for other purposes before the first user data SGPRs.
+//
+// @param conv : The calling convention of the shader
+unsigned PalMetadata::getNumberOfSgprsBeforeUserData(unsigned conv) {
+  // GFX9+ merged shader have an extra 8 SGPRs before user data.
+  if (conv != CallingConv::AMDGPU_VS && m_pipelineState->getTargetInfo().getGfxIpVersion() >= GfxIpVersion{9, 0, 0})
+    return 8;
+  return 0;
+}
+
+// =====================================================================================================================
+// Returns the offset of the userDataMapping from firstUserDataNode.  Returns UINT_MAX if it cannot be found.
+//
+// @param firstUserDataNode : An iterator identifing the starting point for the search.  It is assumed that it points to
+//                            the first user data node for some shader stage.
+// @param userDataMapping : The user data mapping that is being search for.
+//
+unsigned PalMetadata::getOffsetOfUserDataReg(std::map<msgpack::DocNode, msgpack::DocNode>::iterator firstUserDataNode,
```

Comment:
This now scans the user data regs in metadata once for each thing we want to find in it. Probably not much of an issue performance-wise; are you thinking it is more clear?

Reply:
Yes, I did think it was clearer, and I did not think the performance will be significant.  The number of entries that are being searched is limited to 32 entries.